### ISSUE TITLE
Why hyphenate HTTP Headers containing underscores ?

### ISSUE THREAD

**Issue by pamiel** [pintsized/lua-resty-http#114]
Hi all,

`http_header.lua` file has an hyphenate function that replaces the `_` (underscore) of any HTTP headers names (of HTTP requests) into an `-` (hyphen).

While it looks there has been a lot of discussions on whether it is allowed or not to have HTTP header names containing underscores, it looks that it is now quite well accepted, and several tools are relying on specific HTTP headers which names contain underscore. One well known example is the famous `SSL_CLIENT_CERT` header.

The current lua-resty-http implementation does not allow sending HTTP requests with such headers because they are automatically "hyphenated", making the remote server ignore these headers...
Could you please tell me what is the original reason why such hyphenate is done here ? Would it make sense to remove it now ? It is a blocker for me to implement a specific plug-in on Kong API Gateway, using pintsized/lua-resty-http.

Thanks for your help and feedbacks.

**Comment by pintsized** [comment]
The original rationale was compatibility with the ngx_lua behaviour: https://github.com/openresty/lua-nginx-module#ngxreqget_headers

It stems (I guess) from Lua table indexes not accepting hyphens without the `[" "]` syntax, and also how header field names are translated from Nginx `HTTP_` variables: http://nginx.org/en/docs/http/ngx_http_core_module.html#var_http_

Note that this was written before the `raw` mode was added to `ngx.req.get_headers`, which was presumably added for similar reasons.

It's important to have an abstraction for headers, because they must be normalised for case. The mistake was to assume this convention around underscores (widely a constraint in all web servers and also the default behaviour of ngx_lua) to be a sane default.

I *think* it would be safe to change this behaviour, so long as it is documented, as opposed to some kind of option to leave underscores intact. It could break some implementations though.

You're more than welcome to submit a pull request, otherwise I will get to this as soon as I have some time.

**Comment by pamiel** [comment]
> I think it would be safe to change this behaviour, so long as it is documented, as opposed to some kind of option to leave underscores intact. It could break some implementations though.

Sorry, I'm not sure about what you are recommending: do you think it is now safe to remove the hyphenate ?
(I'm confused with the beginning of your sentence that says "it would be safe to change this behavior" => new behavior would be to no longer hyphenate; while the end of your sentence says "as opposed to some kind of option to leave underscores intact" which says that the option to not touch the underscore (i.e. to not hyphenate) is an "opposite" solution).

**Comment by pintsized** [comment]
Sorry, by "option to leave underscores intact" I mean a non-default configuration option, so we retain backwards compatibility by default, but if you need underscored header names (like in your case) you can turn this on. So option 1 is break compatibility, and option 2 is preserve compatibility with a configuration option.

The emphasis was on "*think*" though, in that I'm not sure what the best approach is, and I'm happy for others to influence the decision. Here's how I see it:

The current behaviour is wrong, and the correct behaviour would be case insensitive head fields, with hyphen and underscore both preserved as valid characters. If we change the behaviour, we probably wont break 99% of use cases in reality. What would break is someone using underscore notation to specify header field names, when they wish them to be translated to hyphen. i.e.

```lua
local res, err = httpc:request_uri("http://example.com/", {
    headers = {
          content_type = "application/x-www-form-urlencoded",
    },
})

-- Request header translates to "content-type" over the wire
```

Whilst it might seem odd, someone might write code like that if they're expecting it to work as per `ngx.req.get_headers()` and don't like typing `headers["Content-Type"]` for example, since `headers.content-type` is not legal. I know for sure this wont break any of my code, but it might well break something somewhere I guess, even though it is incorrect.

So option 1 is, we change the behaviour and make a note of it, and if it breaks user code when upgrading lets hope people test first, and update their code.

Option 2 means we have to support both, and we must add a config parameter to switch behaviours. I don't like this, but can be persuaded if the consensus shows that option 1 will break real world things.

It's a weird issue. Technically it's a bug (which should just be fixed), but since it followed convention there is a chance people expect it to work in the broken way. And either way it only changes corner cases (underscored header fields are hardly supported or seen anywhere, and most people will write header field names in the format they wish to use).

TL;DR I think we should just fix it but I have a niggling voice in my head saying "never break backwards compatibility"!

Thoughts?

**Comment by pintsized** [comment]
a1d326de19905203f3aedb2f8f52ceba3c3d3871 This is now "fixed" in master, i.e. we no longer transform underscores to hyphens. I'm not doing a new luarocks / opm release for a little bit whilst I fix some other things, so if anyone has an issue with this behaviour changing, now is your time to shout.

**Comment by pamiel** [comment]
Thanks a lot @pintsized !
Obviously, I have no objection to this change :P

**Comment by pamiel** [comment]
My 2 cents looking to the source code: it may be required to increment the version number in
```
local _M = {
    _VERSION = '0.10',
}
```
;)

**Comment by pintsized** [comment]
Sure, will update version numbers when I do a release.
